### PR TITLE
feat: Add sssd custom settings

### PR DESCRIPTION
**Pull request by brakkio86**
Enhancement: allow sssd custom settings

Reason: sssd.conf is fixed and not costumizable

Result:

Issue Tracker Tickets (Jira or BZ if any):

### CODE REVIEW

**Comment by richm** [comment]
Thanks - looks useful - @justin-stephenson wdyt?

since `ad_integration_sssd_custom_settings` is part of the public interface, it needs the following:
* defined in defaults/main.yml with a default value e.g. `ad_integration_sssd_custom_settings: []`
* described in README.md
* some sort of test for this - specify `ad_integration_sssd_custom_settings`, confirm that they are in sssd.conf

**Comment by justin-stephenson** [comment]
I am good with allowing custom sssd options for this role.

**Comment by brakkio86** [comment]
Hello @richm,
I've added the default section and described in readme. What do you mean with "*some sort of test for this - specify `ad_integration_sssd_custom_settings`, confirm that they are in sssd.conf*"? In my opinion, module `community.general.ini_file` already do that. Or you mean something else?

**Comment by richm** [comment]
> Hello @richm, I've added the default section and described in readme. What do you mean with "_some sort of test for this - specify `ad_integration_sssd_custom_settings`, confirm that they are in sssd.conf_"? In my opinion, module `community.general.ini_file` already do that. Or do you mean something else?

For example - take a look at https://github.com/linux-system-roles/ad_integration/blob/main/tests/tests_dyndns.yml - in fact, IMO you could change this test - add `ad_integration_sssd_custom_settings:` with one setting to `vars:`, and check for it in the `Test - Verify sssd.conf options were written` task.  I realize it may seem like we are testing the workings of the `ini_file` module, but our QE is going to want to ensure that the parameter was indeed written to the config file, and would strongly prefer an automated way to do that, rather than having to manually inspect the resulting file.

**Comment by brakkio86** [comment]
Hello, I've wroted a test. @richm what do you think?

**Comment by brakkio86** [comment]
I've modified the code in order to remove the specified custom parameters. Is it ok now?

File: tasks/main.yml [hunk 1]
```diff
@@ -231,6 +231,20 @@
     - item.value != ''
   notify: Handler for ad_integration to restart services
 
+- name: Configure custom SSSD settings
+  community.general.ini_file:
+    path: /etc/sssd/sssd.conf
+    state: "{{ item.state | default 'present' }}"
```
Comment (richm) [inline tasks/main.yml:237]:
```suggestion
    state: "{{ item.state | default('present') }}"
```

File: README.md [hunk 2]
```diff
@@ -207,6 +207,19 @@ Useful if some specific configuration like --user-principal=host/name@REALM or -
 See man realm for details.
 Example: ad_integration_join_parameters: "--user-principal host/client007@EXAMPLE.COM"
 
+
```
Comment (richm) [inline README.md:210]:
```suggestion
```

File: README.md [hunk 3]
```diff
@@ -207,6 +207,19 @@ Useful if some specific configuration like --user-principal=host/name@REALM or -
...
+#### ad_integration_sssd_custom_settings
+
+A list of custom setting to be included into the [domain/<REALM>] section
```
Comment (richm) [inline README.md:212]:
```suggestion
A list of custom setting to be included into the `[domain/$REALM]` section
```

File: README.md [hunk 4]
```diff
@@ -207,6 +207,19 @@ Useful if some specific configuration like --user-principal=host/name@REALM or -
...
+of the sssd.conf file. See sssd.conf man pages for details.
+
```
Comment (richm) [inline README.md:214]:
```suggestion

```

File: README.md [hunk 5]
```diff
@@ -207,6 +207,19 @@ Useful if some specific configuration like --user-principal=host/name@REALM or -
 See man realm for details.
 Example: ad_integration_join_parameters: "--user-principal host/client007@EXAMPLE.COM"
 
+
+#### ad_integration_sssd_custom_settings
+
+A list of custom setting to be included into the [domain/<REALM>] section
+of the sssd.conf file. See sssd.conf man pages for details.
+ 
+Example:
```
Comment (richm) [inline README.md:215]:
```suggestion
Example:

```

File: README.md [hunk 6]
```diff
@@ -207,6 +207,19 @@ Useful if some specific configuration like --user-principal=host/name@REALM or -
...
+```yaml
+ad_integration_sssd_custom_settings:
+ - key: "configuration_name"
```
Comment (richm) [inline README.md:219]:
```suggestion
  - key: "configuration_name"
```

File: README.md [hunk 7]
```diff
@@ -207,6 +207,19 @@ Useful if some specific configuration like --user-principal=host/name@REALM or -
...
+   value: "configuration_value"
```
Comment (richm) [inline README.md:220]:
```suggestion
    value: "configuration_value"
```

File: README.md [hunk 8]
```diff
@@ -207,6 +207,19 @@ Useful if some specific configuration like --user-principal=host/name@REALM or -
 See man realm for details.
 Example: ad_integration_join_parameters: "--user-principal host/client007@EXAMPLE.COM"
 
+#### ad_integration_sssd_custom_settings
+
+A list of custom setting to be included into the `[domain/$REALM]` section
+of the sssd.conf file. See sssd.conf man pages for details.
+Example:
+
+
```
Comment (richm) [inline README.md]:
```suggestion
```

markdownlint does not like having two blank lines here - just remove one of the blank lines

**Comment by richm** [comment]
[citest]

File: tests/tests_sssd_custom_setttings.yml [hunk 9]
```diff
@@ -0,0 +1,98 @@
+# SPDX-License-Identifier: MIT
+---
+
+- name: Ensure that the role configures dynamic dns
+  hosts: all,!ad
+  gather_facts: false  # test that role works in this case
+  vars:
+    ad_integration_sssd_custom_settings:
```
Comment (richm) [inline tests/tests_sssd_custom_setttings.yml:8]:
```suggestion
    ad_integration_realm: sample-realm.com
    __ad_integration_test_sssd_config_only: true
    ad_integration_sssd_custom_settings:
```

File: tests/tests_sssd_custom_setttings.yml [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+            state: absent
+
+    - name: Test - Re-run the system role to remove vars
+      block:
+        - name: Rerun system role
+          ansible.builtin.include_role:
+            name: linux-system-roles.ad_integration
+          vars:
+            ad_integration_sssd_custom_settings: "{{ update_list }}"
+        - name: Restart sssd
```
Comment (richm) [inline tests/tests_sssd_custom_setttings.yml:74]:
```suggestion
```

File: tests/tests_sssd_custom_setttings.yml [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+          ansible.builtin.service:
```
Comment (richm) [inline tests/tests_sssd_custom_setttings.yml:75]:
```suggestion
```

File: tests/tests_sssd_custom_setttings.yml [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+            name: sssd
```
Comment (richm) [inline tests/tests_sssd_custom_setttings.yml:76]:
```suggestion
```

File: tests/tests_sssd_custom_setttings.yml [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            state: restarted
```
Comment (richm) [inline tests/tests_sssd_custom_setttings.yml:77]:
```suggestion
```

Comment (richm) [reply tests/tests_sssd_custom_setttings.yml:77]: Don't need to restart sssd if we just want to check what's in the config file

**Comment by richm** [comment]
In addition to the other suggested changes, change handlers/main.yml like this:

```
   loop: "{{ __ad_integration_services }}"
+  when: not __ad_integration_test_sssd_config_only | d(false)
```
that way the handler notifications won't actually attempt to restart realmd or sssd when the config files are changed, and we only want to test the config files

**Comment by brakkio86** [comment]
@richm What is the porpouse of the variable `__ad_integration_test_sssd_config_only`?

**Comment by brakkio86** [comment]
> @richm What is the porpouse of the variable `__ad_integration_test_sssd_config_only`?

Ignore it, I found the answer by myself.

**Comment by brakkio86** [comment]
> In addition to the other suggested changes, change handlers/main.yml like this:
> 
> ```
>    loop: "{{ __ad_integration_services }}"
> +  when: not __ad_integration_test_sssd_config_only | d(false)
> ```
> 
> that way the handler notifications won't actually attempt to restart realmd or sssd when the config files are changed, and we only want to test the config files

Done as you suggest.

**Comment by brakkio86** [comment]
> @richm What is the porpouse of the variable `__ad_integration_test_sssd_config_only`?

Ignore it, I found the answer by myself

**Comment by richm** [comment]
[citest]

**Review by richm — APPROVED** [review]

**Comment by richm** [comment]
note - I'm going to merge this even though ansible 2.9 tests are broken - the problem is that the `ini` lookup works differently in 2.9 and later, and 2.9 tries to evaluate the 2.x lookup call, even if excluded by a `when` - so I need to figure out how we will test this with 2.9 (and this applies to the dyndns support as well - I will refactor  that too).

**Comment by richm** [comment]
@brakkio86 can you give me some examples of settings that you need to set in sssd.conf in order for ad_integration to work in your environment?  We would like to add some tests for this feature, but we want to know some settings to apply to sssd.conf

**Pull request merged**